react/ConnectedUIRouter.tsx

```diff
@@ -1,32 +1,43 @@
-import { servicesPlugin, UIRouter, UIRouterProps } from '@uirouter/react';
-import * as PropTypes from 'prop-types';
-import * as React from 'react';
-
-import { createReduxPlugin } from '../core';
-
-export class ConnectedUIRouter extends React.Component<UIRouterProps, any> {
-  reduxPlugin;
-  router;
-
-  static contextTypes = {
-    store: PropTypes.object,
-  };
-
-  constructor(props, context) {
-    super(props, context);
-    this.reduxPlugin = createReduxPlugin(context.store);
-    this.router = props.router;
-    this.router.plugin(servicesPlugin);
-    props.plugins.forEach(plugin => this.router.plugin(plugin));
-    this.router.plugin(this.reduxPlugin);
-    if (props.config) props.config(this.router);
-    (props.states || []).forEach(state =>
-      this.router.stateRegistry.register(state)
+import { servicesPlugin, UIRouter, UIRouterProps } from "@uirouter/react";
+import { ReactReduxContext } from "react-redux";
+import * as React from "react";
+import { useContext, useRef } from "react";
+
+import { createReduxPlugin } from "../core";
+
+interface ConnectedUIRouterProps extends UIRouterProps {
+  children: React.ReactElement;
+}
+
+export function ConnectedUIRouter({
+  children,
+  router: routerFromProps,
+  plugins,
+  config,
+  states,
+}: ConnectedUIRouterProps) {
+  const init = useRef(false);
+
+  const { store } = useContext(ReactReduxContext);
+  const router = useRef(routerFromProps);
+  const reduxPlugin = useRef(createReduxPlugin(store));
+
+  // let's initialise the plugins and set up the router
+  if (init.current !== true) {
```

Comment:
Since you are moving to `Hooks` anyway, maybe you should convert this to `useEffect` to be compliant with the React docs.

Docs of `useEffect` suggest using it for side effects etc. So it would be a good use-case for this code.
It would look like below, with empty array as second parameter to ensure that it fires only once.
```
useEffect(setupRouter, [])
```

https://reactjs.org/docs/hooks-reference.html#useeffect

Reply:
I do not think that is possible as the router would be initialized only **after** the initial render. As for me, the router should be set up synchronously, so the current implementation looks reasonable